src/main/java/org/rascalmpl/maven/AbstractRascalMojo.java

```diff
@@ -0,0 +1,380 @@
+			throw new MojoExecutionException("nested " + mainClass + " was killed", e);
+		}
+		catch (Throwable e) {
+			throw new MojoExecutionException("error launching " + mainClass, e);
+		}
+	}
+
+	/**
+	 * Finds the rascal.jar file that the pom.xml depends on.
+	 * When the current project is rascal itself we resolve to a declared bootstrap
+	 * dependency.
+	 * @throws MojoExecutionException
+	 */
+	protected Path detectedDependentRascalArtifact(Log log, MavenProject project) {
+		try {
+			if (project.getArtifactId().equals("org.rascalmpl:rascal")) {
+				// we are in bootstrap mode and must find a previously released
+				// version to kick-off from
+				return installBootstrapRascalVersion();
+			}
+
+			for (Object o : project.getArtifacts()) {
+				Artifact a = (Artifact) o;
+
+				if (a.getArtifactId().equals("org.rascalmpl:rascal")) {
+					File file = a.getFile().getAbsoluteFile();
+
+					if (a.getSelectedVersion().compareTo(getReferenceRascalVersion()) >= 0) {
+						return file.toPath();
+					}
+					else {
+						log.warn("Rascal version in pom.xml dependency is too old for this Rascal maven plugin. " + getReferenceRascalVersion() + " or later expected.");
+						log.warn("Using a newer version org.rascalmpl:rascal:" + bootstrapRascalVersion + "; please add it to the pom.xml");
+						return installBootstrapRascalVersion();
+					}
+				}
+			}
+
+			log.warn("Pom.xml is missig a dependency on org.rascalmpl:rascal:" + getReferenceRascalVersion() + " (or later).");
+			return null;
+		}
+		catch (OverConstrainedVersionException e) {
+			log.error("Rascal version is over-constrained (impossible to resolve). Expected " + getReferenceRascalVersion() + " or later. Have to abort.");
+			throw new RuntimeException(e);
+		}
+		catch (MojoExecutionException e) {
+			log.error("Unable to just-in-time-install the required (bootstrap) version of Rascal. Have to abort.");
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected List<Path> collectDependentArtifactLibraries(MavenProject project) throws URISyntaxException, IOException {
+		List<Path> libs = new LinkedList<>();
+
+		for (Object o : project.getArtifacts()) {
+			Artifact a = (Artifact) o;
+			File file = a.getFile().getAbsoluteFile();
+
+			libs.add(file.toPath());
+		}
+
+		return libs;
+	}
+
+	protected final ArtifactVersion getReferenceRascalVersion() {
+		return new DefaultArtifactVersion("0.41.0-RC16");
+	}
+
+	protected Path installBootstrapRascalVersion() throws MojoExecutionException {
+		// download the boostrap version from our maven site
+		executeMojo(
+			plugin(
+				"org.apache.maven.plugins",
+				"maven-dependency-plugin",
+				"3.1.1"
+			),
+			goal("get"),
+			configuration(
+      			  element("artifact", "org.rascalmpl:rascal:" + bootstrapRascalVersion)
+    		),
+			executionEnvironment(
+				project,
+				session,
+				pluginManager
+			)
+		);
+
+		// resolve the jar file in the .m2 repository
+		return Path.of(session.getSettings().getLocalRepository(),
```

Comment:
hmm, normally maven deals with that, you give it a coordinate, and it should resolve it? but maybe because we're getting it "outside" of the pom calculation?

Reply:
I wish there was a more "supported" API for this. I'd definitely rewrite to that.